api/include/opentelemetry/trace/propagation/b3_propagator.h

```diff
@@ -0,0 +1,276 @@
+                           context::Context &context) noexcept override
+  {
+    SpanContext span_context = ExtractImpl(getter, carrier);
+    nostd::shared_ptr<Span> sp{new DefaultSpan(span_context)};
+    return context.SetValue(kSpanKey, sp);
+  }
+
+  static SpanContext GetCurrentSpan(const context::Context &context)
+  {
+    context::Context ctx(context);
+    context::ContextValue span = ctx.GetValue(kSpanKey);
+    if (nostd::holds_alternative<nostd::shared_ptr<Span>>(span))
+    {
+      return nostd::get<nostd::shared_ptr<Span>>(span).get()->GetContext();
+    }
+    return SpanContext::GetInvalid();
+  }
+
+  static TraceId GenerateTraceIdFromString(nostd::string_view trace_id)
+  {
+    uint8_t buf[kTraceIdHexStrLength / 2];
+    uint8_t *b_ptr = buf;
+    GenerateBuffFromHexStrPad0(trace_id, sizeof(buf), buf);
+    return TraceId(buf);
+  }
+
+  static SpanId GenerateSpanIdFromString(nostd::string_view span_id)
+  {
+    uint8_t buf[kSpanIdHexStrLength / 2];
+    uint8_t *b_ptr = buf;
+    GenerateBuffFromHexStrPad0(span_id, sizeof(buf), buf);
+    return SpanId(buf);
+  }
+
+  static TraceFlags GenerateTraceFlagsFromString(nostd::string_view trace_flags)
+  {
+    if (trace_flags.length() != 1 || (trace_flags[0] != '1' && trace_flags[0] != 'd'))
+    {  // check for invalid length of flags and treat 'd' as sampled
+      return TraceFlags(0);
+    }
+    return TraceFlags(TraceFlags::kIsSampled);
+  }
+
+private:
+  // Converts hex numbers (string_view) into bytes stored in a buffer and pads buffer with 0.
+  static void GenerateBuffFromHexStrPad0(nostd::string_view hexStr, int bufSize, uint8_t *buf)
+  {  // we are doing this starting from "right" side for left-padding
+    nostd::string_view::size_type posInp = hexStr.length();
+    int posOut                           = bufSize;
+    while (posOut--)
+    {
+      int val = 0;
+      if (posInp)
+      {
+        int hexDigit2 = HexToInt(hexStr[--posInp]);  // low nibble
+        int hexDigit1 = 0;
+        if (posInp)
+        {
+          hexDigit1 = HexToInt(hexStr[--posInp]);
+        }
+        if (hexDigit1 < 0 || hexDigit2 < 0)
+        {  // malformed hex sequence. Fill entire buffer with zeroes.
+          for (int j = 0; j < bufSize; j++)
+          {
+            buf[j] = 0;
+          }
+          return;
+        }
+        val = hexDigit1 * 16 + hexDigit2;
+      }
+      buf[posOut] = val;
+    }
+  }
+
+  // Converts a single character to a corresponding integer (e.g. '1' to 1), return -1
+  // if the character is not a valid number in hex.
+  static int8_t HexToInt(char c)
+  {
+    if (c >= '0' && c <= '9')
+    {
+      return (int)(c - '0');
+    }
+    else if (c >= 'a' && c <= 'f')
+    {
+      return (int)(c - 'a' + 10);
+    }
+    else if (c >= 'A' && c <= 'F')
+    {
+      return (int)(c - 'A' + 10);
+    }
+    else
+    {
+      return -1;
+    }
+  }
+
+  static SpanContext ExtractImpl(Getter getter, const T &carrier)
```

Comment:
Might be a nitpick, but I think we should validate the parsed trace_id, span_id lengths for validity. Say if a malformed header with too large of a trace_id gets sent, currently this logic will produce a valid `SpanContext`, but with some of the trace_id cut off - my opinion is we should just fail early in these cases.

Reply:
Thanks.
Would like to get more opinion on this topic [Robustness](https://en.wikipedia.org/wiki/Robustness_principle) vs [it's potential harmful consequences](https://tools.ietf.org/html/draft-iab-protocol-maintenance-01) from others as I have not found clear guidance on this one.

Reply:
In case of any error in parsing `trace_id` and `span_id`, the right approach should be to set them invalid ( i.e. all zeros ) and return the SpanContext. This will ensure that SpanContext will be invalid.